src/solvers/LevenbergMarquardt.jl

```diff
@@ -0,0 +1,279 @@
+* `jacF` – the Jacobian of ``F``
+* `x` – an initial value ``x ∈ \mathcal M``
+
+# Optional
+* `evaluation` – ([`AllocatingEvaluation`](@ref)) specify whether the gradient works by allocation (default) form `gradF(M, x)`
+  or [`MutatingEvaluation`](@ref) in place, i.e. is of the form `gradF!(M, X, x)`.
+* `retraction_method` – (`default_retraction_method(M)`) a `retraction(M,x,ξ)` to use.
+* `stopping_criterion` – ([`StopWhenAny`](@ref)`(`[`StopAfterIteration`](@ref)`(200), `[`StopWhenGradientNormLess`](@ref)`(10.0^-8))`)
+  a functor inheriting from [`StoppingCriterion`](@ref) indicating when to stop.
+...
+and the ones that are passed to [`decorate_options`](@ref) for decorators.
+
+# Output
+
+the obtained (approximate) minimizer ``x^*``, see [`get_solver_return`](@ref) for details
+
+# References
+
+[^Adachi2022]:
+    > S. Adachi, T. Okuno, and A. Takeda, “Riemannian Levenberg-Marquardt Method with Global
+    > and Local Convergence Properties.” arXiv, Oct. 01, 2022.
+    > doi: [10.48550/arXiv.2210.00253](https://doi.org/10.48550/arXiv.2210.00253).
+[^Peeters1993]:
+    > R. L. M. Peeters, “On a Riemannian version of the Levenberg-Marquardt algorithm,”
+    > VU University Amsterdam, Faculty of Economics, Business Administration and Econometrics,
+    > Serie Research Memoranda 0011, 1993.
+    > link: [https://econpapers.repec.org/paper/vuawpaper/1993-11.htm](https://econpapers.repec.org/paper/vuawpaper/1993-11.htm).
+"""
+function LevenbergMarquardt(
+    M::AbstractManifold, F::TF, gradF::TDF, x; kwargs...
+) where {TF,TDF}
+    x_res = copy(M, x)
+    return LevenbergMarquardt!(M, F, gradF, x_res; kwargs...)
+end
+
+@doc raw"""
+    NonlinearLeastSquaresProblem{T} <: Problem{T}
+
+A type for nonlinear least squares problems.
+`T` is a [`AbstractEvaluationType`](@ref) for the gradient function.
+
+
+specify a problem for gradient based algorithms.
+
+# Fields
+* `M`        – a manifold ``\mathcal M``
+* `F`        – a function ``F: \mathcal M → ℝ^d`` to minimize
+* `jacF!!`   – Jacobian of the function ``F``
+
+Depending on the [`AbstractEvaluationType`](@ref) `T` the gradient has to be provided
+
+* as a function `x -> X` that allocates memory for `X` itself for an [`AllocatingEvaluation`](@ref)
+* as a function `(X,x) -> X` that work in place of `X` for an [`MutatingEvaluation`](@ref)
+
+# Constructors
+    GradientProblem(M, cost, gradient; evaluation=AllocatingEvaluation())
+
+# See also
+[`LevenbergMarquardt`](@ref), [`GradientDescentOptions`](@ref)
+"""
+struct NonlinearLeastSquaresProblem{T,mT<:AbstractManifold,TF,TJ} <: Problem{T}
+    M::mT
+    F::TF
+    jacobian!!::TJ
+end
+function NonlinearLeastSquaresProblem(
+    M::mT, F::TF, jacF::TJ; evaluation::AbstractEvaluationType=AllocatingEvaluation()
+) where {mT<:AbstractManifold,TF,TJ}
+    return NonlinearLeastSquaresProblem{typeof(evaluation),mT,TF,TJ}(M, F, jacF)
+end
+
+function (d::DebugGradient)(::NonlinearLeastSquaresProblem, o::Options, i::Int)
```

Comment:
If you problem is an AbstractGradientProblem this might – for best of cases – work directly without an implementation?

Reply:
I'm still unsure how to model it, and for now I'm focusing more on making it actually work. I will work out the structure when I'm sure it works, OK?

Reply:
That is of course totally fine, Then we keep this as a reminder. I can also check in the end where we can spare a few implementations by inheritance then.